### PR TITLE
net: introduce find_candidate_nics()

### PR DESCRIPTION
**Pull request by cjp256**
find_fallback_nic_on_linux(), etc. provides valuable filtering of
network interfaces in an effort to determine the best candidate
for the fallback interface.

Expose this logic with a new set of methods for finding the candidate
network interfaces.  These methods can be used by data sources which
cannot rely on the fallback interface being the correct choice.

Note that the MAC address filtering is now part of
find_candidate_nics_on_linux().  This should be consistent behavior
as find_fallback_nic_on_linux() never selected an interface without
a MAC.  find_fallback_nic_on_linux() continues to prefer eth0,
but we make no such distinction in the candidate search.

Signed-off-by: Chris Patterson <cpatterson@microsoft.com>

### CODE REVIEW

File: cloudinit/net/__init__.py [hunk 1]
```diff
@@ -449,36 +500,39 @@ def find_fallback_nic_on_linux(blacklist_drivers=None):
...
     # and ignore interfaces that are definitely disconnected
     connected = []
     possibly_connected = []
-    for interface in potential_interfaces:
+    for interface in get_devicelist():
+        if interface == "lo":
+            continue
+        driver = device_driver(interface)
+        if driver in blacklist_drivers:
+            LOG.debug(
```
Comment (cjp256) [inline cloudinit/net/__init__.py:512]: I don't know if these debug logs are excessive, but I figured we should have some indication of why we are ignoring candidate NICs.

**Comment by cjp256** [comment]
@eb3095 This may be of interest to you?

**Comment by eb3095** [comment]
> @eb3095 This may be of interest to you?

Indeed, was hoping it was ready for the last PR, but left a note in a comment about switching out for it when it was available. This solves a few issues for us as well.

**Review by anhvoms — APPROVED** [review]

File: cloudinit/net/__init__.py [hunk 2]
```diff
@@ -389,8 +389,23 @@ def is_disabled_cfg(cfg):
     return cfg.get("config") == "disabled"
 
 
-def find_fallback_nic(blacklist_drivers=None):
-    """Return the name of the 'fallback' network device."""
+def find_candidate_nics(blacklist_drivers: Optional[List] = None) -> List[str]:
```
[the remark this answers is not in the document]
Comment (cjp256) [reply cloudinit/net/__init__.py]: Good point fixed everywhere!

**Review by anhvoms — APPROVED** [review]

**Review by TheRealFalcon — APPROVED** [review]
LGTM! This is really convenient refactor and also simplifying of some of the logic. Thanks!

Wrt to the debug logs, I think it is fine. Better to have a trail than nothing at all.

**Pull request merged**